llama_index/llms/openai.py

```diff
@@ -238,36 +246,75 @@ def gen() -> ChatResponseGen:
                     delta = response.choices[0].delta
                 else:
                     delta = {}
+
+                if response_id is None:
+                    response_id = response.id
+
+                if response_id != response.id:
+                    # start of a new response
+                    is_function = False
+                    response_ix = 0
+
+                # update using deltas
                 role = delta.role or MessageRole.ASSISTANT
                 content_delta = delta.content or ""
                 content += content_delta
 
-                function_call_delta = delta.function_call
-                if function_call_delta is not None:
-                    function_dict = function_call_delta.dict()
-
-                    if function_call is None:
-                        function_call = function_dict
-
-                        ## ensure we do not add a blank function call
-                        if function_call.get("function_name", "") is None:
-                            del function_call["function_name"]
+                # openai provides the delta on one tool-at-a-time
```

Comment:
high-level this is a lot of nested code, can we abstract out into helper functions above?

Reply:
same with the async version

Reply:
good and fair. I have marked these as TODO's in the code -- will pick it up tmr.